middleware/authorization.go

```diff
@@ -1,142 +1,114 @@
 package middleware
 
 import (
-	"context"
 	"fmt"
 	"net/http"
-	"os"
 	"regexp"
-	"time"
 
-	"github.com/RedHatInsights/rbac-client-go"
-	"github.com/RedHatInsights/sources-api-go/config"
 	h "github.com/RedHatInsights/sources-api-go/middleware/headers"
+	"github.com/RedHatInsights/sources-api-go/rbac"
 	"github.com/RedHatInsights/sources-api-go/util"
 	"github.com/labstack/echo/v4"
 	"github.com/redhatinsights/platform-go-middlewares/identity"
 )
 
-var (
-	psks            = config.Get().Psks
-	bypassRbac      = config.Get().BypassRbac
-	rbacClient Rbac = &RbacClient{client: rbac.NewClient(os.Getenv("RBAC_URL"), "sources")}
-)
-
-/*
-Takes the information stored in the context and returns a 401 if we do not
-have authorization to perform "write" things such as POST/PATCH/DELETE.
-
- 1. Checks for PSK (if present) and if it is there and matches any of the
-    PSKs we approve, lets it through.
-
- 2. Sends the x-rh-identity header off to rbac to get an ACL list, and
-    returns whether or not it contains the correct `sources:*:*` permission.
-*/
-func PermissionCheck(next echo.HandlerFunc) echo.HandlerFunc {
-	return func(c echo.Context) error {
-		switch {
-		case bypassRbac:
-			c.Logger().Debugf("Skipping authorization check -- disabled in ENV")
-		case c.Get(h.PSK) != nil:
-			psk, ok := c.Get(h.PSK).(string)
-			if !ok {
-				return fmt.Errorf("error casting psk to string: %v", c.Get(h.PSK))
-			}
-
-			if !pskMatches(psk) {
-				return c.JSON(http.StatusUnauthorized, util.NewErrorDoc("Unauthorized Action: Incorrect PSK", "401"))
-			}
+// PermissionCheck takes the authentication information stored in the context and returns a "401 — Unauthorized" if the
+// given request is not authorized to perform "write" operations such as "POST, PATCH and DELETE".
+//
+//   - When "bypassRbac" is "true", all the requests are authenticated and authorized.
+//   - When using a "psk" in the request, the latter gets authorized if the PSK is in our list of authorized PSKs that
+//     can send requests to Sources.
+//   - Lastly, the requests that come with an "x-rh-identity" header must fulfill one of the following two conditions:
+//   - The request has been authenticated with a certificate, and it's been sent with an allowed "GET", "POST" or
+//     "DELETE" http verb. In the case that it's a "DELETE" request, it will only be authorized to perform that
+//     operation on a subset of paths.
+//   - The request is a regularly authenticated one, so we will call RBAC to verify that the principal that comes in
+//     the header has the authorization to perform the operation in Sources.
+func PermissionCheck(bypassRbac bool, rbacPsks []string, rbacClient rbac.Client) echo.MiddlewareFunc {
```

Comment:
Minor nit-pick (i'm sorry...i can't help it):  The name `rbacPsks` makes me think the psks that are passed in are somehow used with rbac.  However, I think they are just used as a list of allowed psks, right?

Reply:
Very much agreed, yes! I renamed it to `authorizedPsks` so that it its more generic and not tied specifically to RBAC! Thanks for the suggestion!